File: service/src/model/query.go
package model

import (
	"github.com/google/uuid"
)

// COMMON (WEB AND SERVER)

type QueryType int8

const (
	BLOAT QueryType = iota
	ACTIVITY
	REPLICATION
	STATISTIC
	OTHER
)

type QueryCreation string

const (
	Manual QueryCreation = "manual"
	System               = "system"
)

type QueryVariety int8

const (
	DatabaseSensitive QueryVariety = iota
	MasterOnly
	ReplicaRecommended
)

type Query struct {
	Id          uuid.UUID      `json:"id"`
	Name        string         `json:"name"`
	Type        QueryType      `json:"type"`
	Creation    QueryCreation  `json:"creation"`
	Varieties   []QueryVariety `json:"varieties"`
	Params      []string       `json:"params"`
	Description *string        `json:"description"`
	Default     string         `json:"default"`
	Custom      string         `json:"custom"`
	CreatedAt   int64          `json:"createdAt"`
}

type QueryRequest struct {
	Name        string         `json:"name"`
	Type        *QueryType     `json:"type"`
	Description *string        `json:"description"`
	Query       string         `json:"query"`
	Varieties   []QueryVariety `json:"varieties"`
	Params      []string       `json:"params"`
}

type QueryConnection struct {
	Db           Database   `json:"db"`
	Certs        *Certs     `json:"certs"`
	CredentialId *uuid.UUID `json:"credentialId"`
}

type QueryOptions struct {
	Params []any   `json:"params"`
	Limit  *string `json:"limit"`
	Trim   *bool   `json:"trim"`
}

type QueryRunRequest struct {
	Connection   QueryConnection `json:"connection"`
	QueryUuid    *uuid.UUID      `json:"queryUuid"`
	Query        *string         `json:"query"`
	QueryOptions *QueryOptions   `json:"queryOptions"`
}

type QueryKillRequest struct {
	Connection QueryConnection `json:"connection"`
	Pid        int             `json:"pid"`
}

type QueryChartRequest struct {
	Connection QueryConnection `json:"connection"`
	Type       *QueryChartType `json:"type"`
}

type QueryDatabasesRequest struct {
	Connection QueryConnection `json:"connection"`
	Name       string          `json:"name"`
}

type QuerySchemasRequest struct {
	Connection QueryConnection `json:"connection"`
	Name       string          `json:"name"`
}

type QueryTablesRequest struct {
	Connection QueryConnection `json:"connection"`
	Schema     string          `json:"schema"`
	Name       string          `json:"name"`
}

type QueryField struct {
	Name        string `json:"name"`
	DataType    string `json:"dataType"`
	DataTypeOID uint32 `json:"dataTypeOID"`
}

type QueryChartType string

const (
	Databases      QueryChartType = "Databases"
	Connections                   = "Connections"
	DatabaseSize                  = "Database Size"
	DatabaseUptime                = "Database Uptime"
	Schemas                       = "Schemas"
	TablesSize                    = "Tables Size"
	IndexesSize                   = "Indexes Size"
	TotalSize                     = "Total Size"
)

type QueryChart struct {
	Name  string `json:"name"`
	Value any    `json:"value"`
}

type QueryFields struct {
	Fields    []QueryField  `json:"fields"`
	Rows      [][]any       `json:"rows"`
	StartTime int64         `json:"startTime"`
	EndTime   int64         `json:"endTime"`
	Url       string        `json:"url"`
	Options   *QueryOptions `json:"options"`
}

// SPECIFIC (SERVER)

type QueryContext struct {
	Connection QueryConnection
	Session    string
}

type QueryParsed struct {
	LIMIT     int
	UPDATE    int
	DELETE    int
	INSERT    int
	SELECT    int
	FROM      int
	EXPLAIN   int
	Semicolon bool
}

File: web/src/type/query.ts
import {Database} from "./general";
import {Certs} from "./cluster";

// COMMON (WEB AND SERVER)

export enum QueryType {
    BLOAT,
    ACTIVITY,
    REPLICATION,
    STATISTIC,
    OTHER,
    CONSOLE,
}

export enum QueryCreation {
    Manual = "manual",
    System = "system",
}

export enum QueryVariety {
    DatabaseSensitive,
    MasterOnly,
    ReplicaRecommended,
}

export interface Query {
    id: string,
    name: string,
    type: QueryType,
    creation: QueryCreation,
    varieties?: QueryVariety[],
    params?: string[],
    description?: string,
    default: string,
    custom: string,
    createAt: number,
}

export interface QueryRequest {
    type: QueryType,
    query: string,
    name: string,
    description?: string,
    varieties?: QueryVariety[],
    params?: string[],
}

export interface QueryConnection {
    db: Database,
    certs?: Certs,
    credentialId?: string,
}

export interface QueryOptions {
    params?: string[],
    trim?: boolean,
    limit?: string,
}

export interface QueryRunRequest {
    connection: QueryConnection,
    queryUuid?: string,
    query?: string,
    queryOptions?: QueryOptions,
}

export interface QueryKillRequest {
    connection: QueryConnection,
    pid: number,
}

export interface QueryChartRequest {
    connection: QueryConnection,
    type: QueryChartType,
}


export interface QueryDatabasesRequest {
    connection: QueryConnection,
    name: string,
}

export interface QuerySchemasRequest {
    connection: QueryConnection,
    name: string,
}

export interface QueryTablesRequest {
    connection: QueryConnection,
    schema: string,
    name: string,
}

export interface QueryField {
    name: string,
    dataType: string,
    dataTypeOID: number,
}

export enum QueryChartType {
    Databases = "Databases",
    Connections = "Connections",
    DatabaseSize = "Database Size",
    DatabaseUptime = "Database Uptime",
    Schemas = "Schemas",
    TablesSize = "Tables Size",
    IndexesSize = "Indexes Size",
    TotalSize = "Total Size",
}

export interface QueryChart {
    name: string,
    value: any,
}

export interface QueryFields {
    fields: QueryField[],
    rows: any[][],
    url: string,
    startTime: number,
    endTime: number,
    options?: QueryOptions,
}

// SPECIFIC (WEB)
